### PR TITLE
Collapsible panel

### PR DESCRIPTION
**Pull request by kevinluosl**
Updated CollapsiblePanel after first round review by Charles

screen shot 2015-02-11 at 6 07 09 pm

### CODE REVIEW

File: core/selection-controller.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+exports.SelectionController = Montage.specialize(/* @lends SelectionController.prototype */ {
+    constructor: {
+        value: function SelectionController() {
+            this.super();
+        }
+    },
+
+    /**
+     * This is just a placeholder for possible isMultipleSelectionAllowed property
```
Comment (dharcourt) [inline core/selection-controller.js]: nit: Period at the end of the sentence.

File: core/selection-controller.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+     * TODO implement API
+     * @type {boolean}
+     */
+    isMultipleSelectionAllowed: {
+        value: false
+    },
+
+    /**
+     * This is just a placeholder for possible selection property
```
Comment (dharcourt) [inline core/selection-controller.js]: nit: Period at the end of the sentence.

File: ui/collapsible-panel.info/sample/main.reel/main.css [hunk 3]
```diff
@@ -0,0 +1,21 @@
+.Main {
+
+}
```
Comment (dharcourt) [inline ui/collapsible-panel.info/sample/main.reel/main.css]: Any reason why .Main shouldn't be removed?

File: ui/collapsible-panel.info/sample/main.reel/main.html [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                        {"title": "Body 5 title","text": "Repetition content 5","data":5}
+                    ]
+                }
+            },
+            "example2SelectionController2": {
+                "prototype": "montage/core/selection-controller",
+                "properties": {
+                    "isMultipleSelectionAllowed": false,
+                    "selection": [4]
+                }
```
Comment (dharcourt) [inline ui/collapsible-panel.info/sample/main.reel/main.html]: typo: example2SelectionController2 -> example2SelectionController

File: ui/collapsible-panel.info/sample/main.reel/main.html [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        }
+    </script>
+</head>
+<body>
+    <div data-montage-id="main" data-montage-skin="light" class="Main">
+        <h3 style="padding-left: 10px">Example 1: CollapsiblePanel + template-driven content + isMultipleSelectionAllowed = false</h3>
+        <div style="width:500px; margin: 15px">
+            <div data-montage-id="example-1-collapsiblePanel-1" class="Main-CollapsiblePanel">
+                <div data-montage-id="example-1-header-1" data-arg="header">
+                </div>
```
Comment (dharcourt) [inline ui/collapsible-panel.info/sample/main.reel/main.html]: nit: empty <div></div> on one line.

File: ui/collapsible-panel.info/sample/main.reel/main.html [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                <div data-montage-id="example-1-body-1" data-arg="body"></div>
+            </div>
+            <div data-montage-id="example-1-collapsiblePanel-2" class="Main-CollapsiblePanel">
+                <div data-montage-id="example-1-header-2" data-arg="header">
+                </div>
```
Comment (dharcourt) [inline ui/collapsible-panel.info/sample/main.reel/main.html]: nit: empty `<div></div>` on one line.

File: ui/collapsible-panel.info/sample/main.reel/main.html [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+                    <input placeholder="A native HTML input">
+                </div>
+            </div>
+        </div>
+        <h3 style="padding-left: 10px">Example 2: CollapsiblePanel + data-driven repetition + isMultipleSelectionAllowed = false</h3>
+        <div style="width:500px; margin: 15px">
+            <div data-montage-id="example-2-repetition">
+                <div data-montage-id="example-2-collapsiblePanel" class="Main-CollapsiblePanel">
+                    <div data-montage-id="example-2-header" data-arg="header">
+                    </div>
```
Comment (dharcourt) [inline ui/collapsible-panel.info/sample/main.reel/main.html]: nit: empty `<div></div>` on one line.

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            //this.needsDraw = (this._isExpanded != value);
+        },
+        get: function () {
+            return this._isExpanded;
+        }
+    },
+
+    prepareForActivationEvents: {
+        value: function () {
+            var transitionEndListener = {};
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: typo: This is gone :-).

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            var bodyWrapper = this.element.children[1];
+            this._pressComposer.addEventListener("press", this, false);
+            //Listen css transition call back.
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: spelling & spacing: // Listen to CSS transition call back.

File: ui/collapsible-panel.reel/collapsible-panel.js (hunk 8 above)
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: var handler = this.handleTransitionEnd.bind(this);

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            bodyWrapper.addEventListener('webkitTransitionEnd', this.handleTransitionEnd);
+            bodyWrapper.addEventListener('transitionend', this.handleTransitionEnd);
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: bodyWrapper.addEventListener('transitionend', handler);

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    },
+
+    handleTransitionEnd: {
+        value: function () {
+            this.component.parentComponent._isTransition = false;
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: this. _isTransition = false;

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    },
+
+    /**
+     * Do expand or collapse when click on the header.
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: nit: spacing.

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+     * @method
+     * @private
+     */
+    handlePress: {
+        value: function (event) {
+            //Only trigger css animation when clicking on header, this may need change base on how SelectionController handle
+            //multipleSelection and singleSelection.
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: nit: "// " (space)

File: ui/collapsible-panel.reel/collapsible-panel.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            this.isExpanded = !this.isExpanded;
+        }
+    },
+
+    enterDocument: {
+        value: function (firstTime) {
+            if (firstTime) {
+                var header = this.element.children[0];
+                this._pressComposer = new PressComposer();
+                //Register composer here and add event in prepareForActivationEvents.
```
Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: nit: "// " (space)

File: ui/collapsible-panel.reel/collapsible-panel.js (hunk 10 above)
Comment (marchant) [inline ui/collapsible-panel.reel/collapsible-panel.js]: addEventListener is also supposed to have 3 arguments

Comment (dharcourt) [inline ui/collapsible-panel.reel/collapsible-panel.js]: Ah, yes, the 3rd argument is optional in theory but not in practice: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget.addEventListener

Should we actually be passing a true value for useCapture in this case?

**Pull request closed without merging**